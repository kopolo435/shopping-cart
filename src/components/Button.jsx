import React from "react";
import PropTypes from "prop-types";

function Button({ children, onClick, className, label, type }) {
  return label ? (
    <button
      type={type === "button" ? "button" : "submit"}
      onClick={onClick}
      className={className}
      aria-label={label}
    >
      {children}
    </button>
  ) : (
    <button
      type={type === "button" ? "button" : "submit"}
      onClick={onClick}
      className={className}
    >
      {children}
    </button>
  );
}

Button.propTypes = {
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string.isRequired,
  label: PropTypes.string,
  type: PropTypes.string.isRequired,
};

Button.defaultProps = {
  label: null,
  children: "",
};

export default Button;
