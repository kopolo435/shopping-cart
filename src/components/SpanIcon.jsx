import React from "react";
import PropTypes from "prop-types";

function SpanIcon({ text, iconName, className }) {
  return text ? (
    <span>
      {text}
      <span className={className}>{iconName}</span>
    </span>
  ) : (
    <span className={className}>{iconName}</span>
  );
}

SpanIcon.propTypes = {
  text: PropTypes.string,
  iconName: PropTypes.string.isRequired,
  className: PropTypes.string,
};
SpanIcon.defaultProps = {
  text: null,
  className: "material-symbols-outlined",
};

export default SpanIcon;
