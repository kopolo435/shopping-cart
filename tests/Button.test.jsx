import { render, screen } from "@testing-library/react";
import { expect, it, vi } from "vitest";
import React from "react";
import userEvent from "@testing-library/user-event";
import Button from "../src/components/Button";
import SpanIcon from "../src/components/SpanIcon";

it("Renders correct button without label", () => {
  const mockFn = vi.fn(() => {});
  const content = "A単adir al carro";
  const btnClass = "primaryBtn";

  const { container } = render(
    <Button
      type="button"
      content={content}
      onClick={mockFn}
      className={btnClass}
    />
  );

  expect(container).toMatchSnapshot();
});

it("Renders correct button with accesibility label", () => {
  const mockFn = vi.fn(() => {});
  const content = <SpanIcon iconName="add" />;
  const btnClass = "primaryBtn";
  const label = "A単adir al carro";

  const { container } = render(
    <Button
      type="button"
      content={content}
      onClick={mockFn}
      className={btnClass}
      label={label}
    />
  );

  expect(container).toMatchSnapshot();
});

it("Calls onClick function when clicked", async () => {
  const onClick = vi.fn(() => {});
  const content = "A単adir al carro";
  const btnClass = "primaryBtn";
  const user = userEvent.setup();

  render(
    <Button
      type="button"
      content={content}
      onClick={onClick}
      className={btnClass}
    />
  );

  const button = screen.getByRole("button", { name: "A単adir al carro" });

  await user.click(button);
  expect(onClick).toHaveBeenCalled();
});
