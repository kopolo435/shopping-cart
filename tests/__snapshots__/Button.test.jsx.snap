// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders correct button with accesibility label 1`] = `
<div>
  <button
    aria-label="Añadir al carro"
    class="primaryBtn"
    type="button"
  >
    <span
      class="material-symbols-outlined"
    >
      add
    </span>
  </button>
</div>
`;

exports[`Renders correct button without label 1`] = `
<div>
  <button
    class="primaryBtn"
    type="button"
  >
    Añadir al carro
  </button>
</div>
`;
